### PR TITLE
Feature/log full access stack

### PR DESCRIPTION
**Pull request by egreer**
* adding full stack logging when accessing a method for the first time after an allow_phi https://github.com/apsislabs/phi_attrs/issues/15
* Updating the disallow message to include which access was removed
* matching access log message to instance

### CODE REVIEW

File: lib/phi_attrs/phi_record.rb [hunk 1]
```diff
@@ -356,6 +358,28 @@ def class_phi_context
...
     end
 
+    # The unique identifiers for everything with access allowed on this instance.
+    #
+    # @private
+    #
+    # @return Array[String] all the user_id's passed in to allow_phi!
+    #
+    def all_phi_allowed_by
+      all_phi_context.map { |c| c[:user_id] }.to_s
```
Comment (Crisfole) [inline lib/phi_attrs/phi_record.rb]: this says it's supposed to return an `Array[String]`, but it looks like it actually returns a String (of an array). Did you mean:

```suggestion
  all_phi_context.map { |c| c[:user_id].to_s }
```

Comment (egreer) [reply lib/phi_attrs/phi_record.rb]: No should be a string, I pushed up a change with some additional formatting fixes to the string

File: lib/phi_attrs/phi_record.rb [hunk 2]
```diff
@@ -398,9 +422,10 @@ def phi_wrap_method(method_name)
         PhiAttrs::Logger.tagged(*phi_log_keys) do
           raise PhiAttrs::Exceptions::PhiAccessException, "Attempted PHI access for #{self.class.name} #{@__phi_user_id}" unless phi_allowed?
 
-          unless phi_context[:logged]
-            PhiAttrs::Logger.info("'#{phi_allowed_by}' accessing #{self.class.name}. Triggered by method: #{method_name}")
-            phi_context[:logged] = true
+          unless all_phi_context_logged?
+            PhiAttrs::Logger.info("#{self.class.name} access by [#{all_phi_allowed_by}]. Triggered by method: #{method_name}")
+            # phi_context[:logged] = true
```
Comment (HenryKeiter) [inline lib/phi_attrs/phi_record.rb]: Don't need this

File: spec/phi_attrs/logger_spec.rb [hunk 3]
```diff
@@ -109,6 +126,104 @@
           end # Logged Disallowed
         end # Logged Disallowed
       end
+
+      it 'multiple times for nested allows and disallows' do |t|
+        PatientInfo.allow_phi!(file_name + + '1', t.full_description)
```
Comment (HenryKeiter) [inline spec/phi_attrs/logger_spec.rb]: This seems like a syntax error

File: lib/phi_attrs/phi_record.rb [hunk 4]
```diff
@@ -362,10 +362,10 @@ def class_phi_context
     #
     # @private
     #
-    # @return Array[String] all the user_id's passed in to allow_phi!
+    # @return String of all the user_id's passed in to allow_phi!
```
Comment (Crisfole) [inline lib/phi_attrs/phi_record.rb]: 👏

**Pull request merged**